### PR TITLE
Fix integer overflow in WebPImage::getHeaderOffset

### LINKED ISSUE #960: buffer overflow in exiv2/src/webpimage.cpp
**Describe the bug**
I found a buffer overflow bugs in exiv2/src/webpimage.cpp:837 due to integer overflow.
**To Reproduce**
Steps to reproduce the behaviour:
exiv2/build/bin/exiv2 exiv2_getHeaderOffset.poc

**Expected behavior**
In Exiv2::WebPImage::getHeaderOffset, integer overflow occurs when data_size is less than header_size, and it will cause buffer overflow in '&data[i]'.
image
```
#0  0x00007ffff784ce88 in Exiv2::WebPImage::getHeaderOffset (this=<optimized out>, data=0x71a2e0 "\300\\\310\366\377\177", data_size=0x0, header_size=0x4, header=<optimized out>) at /data/Exiv2_AFL/exiv2/src/webpimage.cpp:837
#1  Exiv2::WebPImage::decodeChunks (this=<optimized out>, filesize=<optimized out>) at /data/Exiv2_AFL/exiv2/src/webpimage.cpp:632
#2  0x00007ffff784b2e0 in Exiv2::WebPImage::readMetadata (this=0x719820) at /data/Exiv2_AFL/exiv2/src/webpimage.cpp:508
#3  0x000000000049d59d in Action::Print::printSummary (this=0x71a950) at /data/Exiv2_AFL/exiv2/src/actions.cpp:260
#4  0x000000000049cf30 in Action::Print::run (this=0x71a950, Python Exception <class 'gdb.error'> There is no member named _M_dataplus.:
path=) at /data/Exiv2_AFL/exiv2/src/actions.cpp:215
#5  0x00000000004094e6 in main (argc=<optimized out>, argc@entry=0x2, argv=<optimized out>, argv@entry=0x7fffffffe6c8) at /data/Exiv2_AFL/exiv2/src/exiv2.cpp:77
#6  0x00007ffff68bbb97 in __libc_start_main (main=0x408b30 <main(int, char* const*)>, argc=0x2, argv=0x7fffffffe6c8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe6b8) at ../csu/libc-start.c:310
#7  0x0000000000408a6a in _start ()

   0x7ffff784ce77 <Exiv2::WebPImage::decodeChunks(long)+4279>:	mov    rcx,QWORD PTR [rsp+0x8]
   0x7ffff784ce7c <Exiv2::WebPImage::decodeChunks(long)+4284>:	mov    rdx,QWORD PTR [rsp]
   0x7ffff784ce80 <Exiv2::WebPImage::decodeChunks(long)+4288>:	lea    rsp,[rsp+0x98]
=> 0x7ffff784ce88 <Exiv2::WebPImage::decodeChunks(long)+4296>:	cmp    DWORD PTR [rcx+rsi*1],0xe1ff01ff
   0x7ffff784ce8f <Exiv2::WebPImage::decodeChunks(long)+4303>:	je     0x7ffff784d858 <Exiv2::WebPImage::decodeChunks(long)+6808>
   0x7ffff784ce95 <Exiv2::WebPImage::decodeChunks(long)+4309>:	nop    DWORD PTR [rax]
   0x7ffff784ce98 <Exiv2::WebPImage::decodeChunks(long)+4312>:	lea    rsp,[rsp-0x98]
   0x7ffff784cea0 <Exiv2::WebPImage::decodeChunks(long)+4320>:	mov    QWORD PTR [rsp],rdx


RCX: 0x71a2e0
RSI: 0xdd1d

```
The poc is here:
Just do 'exiv2/build/bin/exiv2 exiv2_getHeaderOffset.poc'

[exiv2_getHeaderOffset.poc.zip](https://github.com/Exiv2/exiv2/files/3390730/exiv2_getHeaderOffset.poc.zip)

### PR DESCRIPTION
**Pull request by piponazo**
This PR fixes #960

### CODE REVIEW

**Review by clanmills — DISMISSED** [review]
Thanks for dealing with this @piponazo.  I set the target == 0.27.2 and label "to master"

**Comment by D4N** [comment]
Since the poc results in a crash on Master, shouldn't we rather fix it on master first?

File: src/webpimage.cpp [hunk 1]
```diff
@@ -827,8 +827,9 @@ namespace Exiv2 {
         }
     }
 
-    long WebPImage::getHeaderOffset(byte *data, long data_size,
-                                    byte *header, long header_size) {
+    long WebPImage::getHeaderOffset(byte* data, long data_size, byte* header, long header_size)
+    {
+        enforce(data_size >= header_size, kerFailedToReadImageData);
```
Comment (D4N) [inline src/webpimage.cpp]: How about this instead?

```suggestion
        if (data_size < header_size) { return -1; }
```

Solves the issue on master and does not require throwing another exception.

Comment (clanmills) [reply src/webpimage.cpp]: Dan:  You are one really smart cookie.  You have a built-in sense for elegance.  And great at thinking about how something might fail to work.

Luis:  You're a really smart cookie as well of course.

**Comment by piponazo** [comment]
> Since the poc results in a crash on Master, shouldn't we rather fix it on master first?

Right. I tested directly on `0.27-maintenance` and I could not reproduce the crash, but on master I could. The only difference is that we end up throwing different exceptions after applying the fix in different branches. 

I propose to merge this as it is in `0.27-maintenance` and then I will take care of adapting the test code in master when mergify do its job.

**Comment by clanmills** [comment]
This sounds good, Luis.  It'd be good to get these things into v0.27.2.  However, we should wrap up RC3 in the next few days and get v0.27.2 out by 2019-07-31.

I made a start on the book today.  At the moment, it's just a framework.  I'll include the first draft in v0.27.3 with a couple of chapters written, and brief notes in the other chapters.

**Review by D4N — APPROVED** [review]

**Pull request merged**